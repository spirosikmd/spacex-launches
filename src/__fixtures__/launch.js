export function createLaunchResponse(data) {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data) {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
