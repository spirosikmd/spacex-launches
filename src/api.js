function sanitizeLaunchesResponse(launches) {
  return launches.map(launch => ({
    flightNumber: launch.flight_number,
    utcDate: new Date(launch.launch_date_utc),
    isSuccessful: launch.launch_success === true && launch.upcoming === false,
    isFailed: launch.launch_success === false && launch.upcoming === false,
    isUpcoming: launch.upcoming === true,
    isTentative: launch.is_tentative === true,
    details: launch.details,
    missionName: launch.mission_name,
    missionIds: launch.mission_id || [],
  }));
}

export async function getLaunches({ sortOrder, sortField, filter }) {
  const response = await fetch(
    `https://api.spacexdata.com/v3/launches?filter=${filter}&sort=${sortField}&order=${sortOrder}`,
    {
      headers: {
        'Content-Type': 'application/json',
      },
    }
  );
  const launches = await response.json();
  return sanitizeLaunchesResponse(launches);
}
