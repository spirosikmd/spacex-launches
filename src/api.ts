import { FLIGHT_NUMBER_FIELD, UTC_DATE_FIELD } from './constants';

const BASE = 'https://api.spacexdata.com/v3';
export const LAUNCHES_BASE = `${BASE}/launches`;
export const FILTER =
  'flight_number,launch_date_utc,launch_success,upcoming,is_tentative,details,mission_name,mission_id';

export interface LaunchData {
  flightNumber: number;
  missionName: string;
  utcDate: Date;
  isSuccessful: boolean;
  isFailed: boolean;
  isUpcoming: boolean;
  isInProgress: boolean;
  isTentative: boolean;
  details: string;
  missionIds: string[];
  missionPatch: string;
}

export interface LaunchResponseData {
  flight_number: number;
  mission_name: string;
  launch_date_utc: string;
  launch_success: boolean;
  upcoming: boolean;
  is_tentative: boolean;
  details: string;
  mission_id: string[];
  links: {
    mission_patch: string;
  };
}

function sanitizeLaunchResponse(launch: LaunchResponseData) {
  return {
    flightNumber: launch.flight_number,
    utcDate: new Date(launch.launch_date_utc),
    isSuccessful: launch.launch_success === true && launch.upcoming === false,
    isFailed: launch.launch_success === false && launch.upcoming === false,
    isUpcoming: launch.launch_success === null && launch.upcoming === true,
    isInProgress: launch.launch_success === null && launch.upcoming === false,
    isTentative: launch.is_tentative === true,
    details: launch.details,
    missionName: launch.mission_name,
    missionIds: launch.mission_id || [],
    missionPatch: (launch.links && launch.links.mission_patch) || '',
  };
}

function sanitizeLaunchesResponse(launches: LaunchResponseData[]) {
  return launches.map(sanitizeLaunchResponse);
}

function getHeaders() {
  return {
    'Content-Type': 'application/json',
  };
}

export function mapSortField(sortField: string) {
  switch (sortField) {
    case FLIGHT_NUMBER_FIELD:
      return 'flight_number';
    case UTC_DATE_FIELD:
    default:
      return 'launch_date_utc';
  }
}

async function handleResponse(response: Response) {
  if (!response.ok) {
    const { error } = await response.json();
    throw new Error(error);
  }

  return response.json();
}

export async function getLaunches({
  sortOrder,
  sortField,
}: {
  sortOrder: string;
  sortField: string;
}) {
  const response = await fetch(
    `${LAUNCHES_BASE}?filter=${FILTER}&sort=${mapSortField(
      sortField
    )}&order=${sortOrder}`,
    {
      headers: getHeaders(),
    }
  );

  const launches = await handleResponse(response);

  return sanitizeLaunchesResponse(launches);
}

export async function getLaunch({ flightNumber }: { flightNumber: number }) {
  const response = await fetch(`${LAUNCHES_BASE}/${flightNumber}`, {
    headers: getHeaders(),
  });

  const launch = await handleResponse(response);

  return sanitizeLaunchResponse(launch);
}
