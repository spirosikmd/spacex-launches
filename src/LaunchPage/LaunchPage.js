import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getLaunch } from '../api';
import Loader from '../Loader';

export class LaunchPage extends Component {
  state = {
    error: null,
    launch: null,
    isLoadingLaunch: true,
  };

  async componentDidMount() {
    try {
      const { flightNumber } = this.props;
      const launch = await getLaunch({ flightNumber });
      this.setState({ isLoadingLaunch: false, launch });
    } catch (error) {
      this.setState({ error, isLoadingLaunch: false });
    }
  }

  render() {
    const { launch, isLoadingLaunch } = this.state;

    if (isLoadingLaunch) {
      return <Loader />;
    }

    return (
      <>
        <div>{launch.missionName}</div>
        <div>{launch.details}</div>
      </>
    );
  }
}

LaunchPage.propTypes = {
  flightNumber: PropTypes.string.isRequired,
};

export default LaunchPage;
