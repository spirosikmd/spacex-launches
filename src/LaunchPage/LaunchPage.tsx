import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  withStyles,
  Theme,
  createStyles,
  WithStyles,
} from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { Redirect, RouteComponentProps } from '@reach/router';
import { getLaunch, LaunchData } from '../api';
import Loader from '../Loader';
import LaunchDateTime from '../LaunchDateTime';
import LaunchStatus from '../LaunchStatus';
import LaunchMissionIds from '../LaunchMissionIds';

const styles = (theme: Theme) =>
  createStyles({
    headline: {
      display: 'flex',
      alignItems: 'center',
      marginBottom: theme.spacing(3),
      [theme.breakpoints.down('xs')]: {
        flexDirection: 'column',
      },
    },
    missionPatch: {
      width: '120px',
      height: '120px',
      marginRight: theme.spacing(3),
      [theme.breakpoints.down('xs')]: {
        marginRight: 0,
        marginBottom: theme.spacing(3),
      },
    },
    missionName: {
      [theme.breakpoints.down('xs')]: {
        textAlign: 'center',
      },
    },
    details: {
      paddingLeft: theme.spacing(1),
      paddingRight: theme.spacing(1),
    },
    date: {
      marginBottom: theme.spacing(1),
    },
    statusMissionIds: {
      display: 'flex',
      alignItems: 'center',
      marginBottom: theme.spacing(2),
    },
    missionIds: {
      marginLeft: theme.spacing(1),
    },
  });

interface LaunchPageProps
  extends RouteComponentProps<{ flightNumber: string }>,
    WithStyles<typeof styles> {}

interface LaunchPageState {
  error: Error | null;
  launch: LaunchData | null;
  isLoadingLaunch: boolean;
}

class LaunchPage extends PureComponent<LaunchPageProps, LaunchPageState> {
  static propTypes = {
    flightNumber: PropTypes.string,
    classes: PropTypes.shape({
      headline: PropTypes.string.isRequired,
      missionPatch: PropTypes.string.isRequired,
      missionName: PropTypes.string.isRequired,
      details: PropTypes.string.isRequired,
      date: PropTypes.string.isRequired,
      statusMissionIds: PropTypes.string.isRequired,
      missionIds: PropTypes.string.isRequired,
    }).isRequired,
  };

  state: LaunchPageState = {
    error: null,
    launch: null,
    isLoadingLaunch: true,
  };

  async componentDidMount() {
    try {
      const { flightNumber } = this.props;
      if (!flightNumber) return;
      const launch = await getLaunch({ flightNumber });
      this.setState({ isLoadingLaunch: false, launch });
    } catch (error) {
      this.setState({ error, isLoadingLaunch: false });
    }
  }

  render() {
    const { classes } = this.props;
    const { launch, isLoadingLaunch, error } = this.state;

    if (isLoadingLaunch) {
      return <Loader />;
    }

    if (error) {
      return <Typography>{error.message}</Typography>;
    }

    if (!launch || launch.isTentative) {
      return <Redirect to="/" noThrow />;
    }

    return (
      <div>
        <div className={classes.headline}>
          {launch.missionPatch && (
            <img
              className={classes.missionPatch}
              src={launch.missionPatch}
              alt={launch.missionName}
            />
          )}
          <Typography variant="h4" className={classes.missionName}>
            {launch.missionName}
          </Typography>
        </div>
        <Typography variant="subtitle1" className={classes.date}>
          <LaunchDateTime
            utcDate={launch.utcDate}
            isTentative={launch.isTentative}
          />
        </Typography>
        <div className={classes.statusMissionIds}>
          <LaunchStatus
            isSuccessful={launch.isSuccessful}
            isFailed={launch.isFailed}
            isUpcoming={launch.isUpcoming}
            isInProgress={launch.isInProgress}
          />
          {launch.missionIds.length > 0 && (
            <div className={classes.missionIds}>
              <LaunchMissionIds missionIds={launch.missionIds} />
            </div>
          )}
        </div>
        <Typography>{launch.details}</Typography>
      </div>
    );
  }
}

export default withStyles(styles)(LaunchPage);
