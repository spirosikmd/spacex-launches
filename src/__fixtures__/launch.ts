import { LaunchResponseData, LaunchData } from '../api';

export function createLaunchResponse(
  data: LaunchResponseData
): LaunchResponseData {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data: LaunchData): LaunchData {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
