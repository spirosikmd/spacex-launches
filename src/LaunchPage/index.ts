import LaunchPage from './LaunchPage';

export default LaunchPage;
