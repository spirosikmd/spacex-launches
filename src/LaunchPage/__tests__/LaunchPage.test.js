import LaunchPage from '../LaunchPage';
import { getLaunch } from '../../api';
import { createLaunch } from '../../__fixtures__/launch';

jest.mock('../../api', () => ({
  getLaunch: jest.fn(),
}));

describe('LaunchPage', () => {
  let props;

  beforeEach(() => {
    props = {
      classes: {
        headline: 'headline',
        missionPatch: 'missionPatch',
        missionName: 'missionName',
        details: 'details',
        date: 'date',
        status: 'status',
      },
      flightNumber: '1',
    };
    const launch = createLaunch();
    getLaunch.mockReturnValue(Promise.resolve(launch));
  });

  it('renders', async () => {
    const launchPage = shallow(LaunchPage, props);
    await Promise.resolve();
    launchPage.update();
    expect(launchPage).toMatchSnapshot();
  });

  it('does not render mission patch when there is no mission patch', async () => {
    getLaunch.mockReturnValue(
      Promise.resolve(createLaunch({ missionPatch: '' }))
    );
    const launchPage = shallow(LaunchPage, props);
    await Promise.resolve();
    launchPage.update();
    expect(launchPage).toMatchSnapshot();
  });

  it('renders a loader when is loading a launch', () => {
    expect(shallow(LaunchPage, props)).toMatchSnapshot();
  });

  it('renders error message when there is an error', async () => {
    getLaunch.mockReturnValue(Promise.reject({ message: 'Not Found' }));
    const launchPage = shallow(LaunchPage, props);
    await Promise.resolve();
    launchPage.update();
    expect(launchPage).toMatchSnapshot();
  });
});
