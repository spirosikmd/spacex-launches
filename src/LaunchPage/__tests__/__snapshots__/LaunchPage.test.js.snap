// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchPage does not render mission patch when there is no mission patch 1`] = `
<div>
  <div
    className="headline"
  >
    <WithStyles(Typography)
      className="missionName"
      variant="h4"
    >
      Test
    </WithStyles(Typography)>
  </div>
  <WithStyles(Typography)
    className="date"
    variant="subtitle1"
  >
    <LaunchDateTime
      isTentative={false}
      utcDate={
        Object {
          "getDate": [MockFunction],
          "getFullYear": [MockFunction],
          "getMonth": [MockFunction],
          "toUTCString": [MockFunction],
        }
      }
    />
  </WithStyles(Typography)>
  <WithStyles(Typography)>
    Details
  </WithStyles(Typography)>
</div>
`;

exports[`LaunchPage renders 1`] = `
<div>
  <div
    className="headline"
  >
    <img
      alt="Test"
      className="missionPatch"
      src="missionPatch"
    />
    <WithStyles(Typography)
      className="missionName"
      variant="h4"
    >
      Test
    </WithStyles(Typography)>
  </div>
  <WithStyles(Typography)
    className="date"
    variant="subtitle1"
  >
    <LaunchDateTime
      isTentative={false}
      utcDate={
        Object {
          "getDate": [MockFunction],
          "getFullYear": [MockFunction],
          "getMonth": [MockFunction],
          "toUTCString": [MockFunction],
        }
      }
    />
  </WithStyles(Typography)>
  <WithStyles(Typography)>
    Details
  </WithStyles(Typography)>
</div>
`;

exports[`LaunchPage renders a loader when is loading a launch 1`] = `<WithStyles(Loader) />`;
